Player warnings limited to once per minute.

FuzTheCat opened the pull request:
Once per minute limitation was automatically written by ChatGPT during a chat session.

FuzTheCat committed:
Update danger_ore.lua

Check against listed ores only so can now build on oil resources. Also updated description for pipes on ores.

FuzTheCat committed:
Update danger_ores.lua

Added pipe & pipe-to-ground to allowed_entities list

FuzTheCat committed:
Update danger_ore.lua

FuzTheCat committed:
Merge branch 'Refactorio:develop' into develop

FuzTheCat committed:
Update danger_ores.lua

Added wooden-chest to allowed entities.

FuzTheCat committed:
Update danger_ores.lua

Change ore_pattern back to tendrils_impure

FuzTheCat committed:
Merge branch 'Refactorio:develop' into develop

FuzTheCat committed:
BnB Feature

FuzTheCat committed:
BnB Preset

FuzTheCat committed:
Fix warnings in belts_n_bullets.lua

grilledham committed:
Add scenario template.

FuzTheCat committed:
Merge branch 'Refactorio:develop' into develop

FuzTheCat committed:
Limit warnings to once per minute.

I had no idea how to do this, so just submitted this file text to ChatGPT, and asked it to adjust the warning to once per minute per player and voila!

grilledham commented on map_gen/maps/diggy/feature/danger_ore.lua:
```diff
@@ -29,10 +29,20 @@ local function banned_entities(allowed_entities)
 
     --- Warning for players when their entities are destroyed
+    --- Note: Edit to limit warning once per minute per player produced completely automatically by ChatGPT
+    local last_warning_time = {}
```
This would desync in multiplayer, it needs to go inside global. See `player_collapses` in https://github.com/Refactorio/RedMew/blob/develop/map_gen/maps/diggy/feature/antigrief.lua for an example.

FuzTheCat replied:
Thank you. I'll see what I can do about getting local variables registered globally.

grilledham commented on map_gen/maps/diggy/feature/danger_ore.lua:
```diff
@@ -29,10 +29,20 @@ local function banned_entities(allowed_entities)
         local p = event.player
         if p and p.valid then
-            p.print('You cannot build that on top of ores, only belts, mining drills, and power poles are allowed.')
+            if not last_warning_time[p.index] then
+                last_warning_time[p.index] = 0
+            end
+            local current_time = game.tick
+            if current_time > last_warning_time[p.index] + (60 * 60) then  -- 60 seconds * 60 ticks per second
+                last_warning_time[p.index] = current_time
+                p.print('You cannot build that on top of ores, only belts, mining drills, and power poles are allowed.')
+            end
```
Minor but I think doing it that way would mean you don't get the warning in the first minute of gameplay.
I'd write it like this (untested):
```suggestion
            local current_time = game.tick
            local player_last_warning_time = last_warning_time[p.index]
            if player_last_warning_time ~= nil and current_time < player_last_warning_time + (60 * 60) then  -- 60 seconds * 60 ticks per second
                return
            end            
            last_warning_time[p.index] = current_time
            p.print('You cannot build that on top of ores, only belts, mining drills, and power poles are allowed.')
           
```

FuzTheCat replied:
Yes, there would be no initial warning for the first minute. It could almost take a minute for the first ore to be uncovered also. But it's good to have the code doing the right thing, thank you.

FuzTheCat replied:
OK, I've done some more edits now and tested locally in multiplayer with two accounts. Does it look good now?

grilledham reviewed (changes requested):
Not bad for a chat bot 😂. But I guess it needs more context to know about Factorio desyncs.

FuzTheCat committed:
Added warning_timeout to the config

FuzTheCat committed:
Made multiplayer desync safe

last_warning_time is now registered globally
Also, use warning_timeout from the config file.
